dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68432 dbr quant overhead[9/x]: precalculate when to skip op_convert_after_hook
* #68431 dbr quant: remove more asserts from hot paths
* #68404 speed up quantized relu6 inplace kernel
* #68374 dbr quant: simplify relatedness logic
* #68373 dbr quant overhead[8/x]: small speedup in op_needs_quantization
* **#68372 dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call**
* #68371 dbr quant overhead[6/x]: remove unneeded isinstance checks in `op_convert_before_hook`
* #68370 dbr quant overhead[5/x]: remove unnecessary asserts
* #68369 dbr quant: clean up key types in AutoQuantizationState mappings
* #68351 dbr quant overhead[4/x]: speed up hook type calculations
* #68350 dbr quant overhead[3/x]: speed up AutoQuantizationState.mark_cur_op_complete

Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

Differential Revision: [D32463759](https://our.internmc.facebook.com/intern/diff/D32463759)

vkuzo committed:
dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call

Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

[ghstack-poisoned]

vkuzo committed:
Update on "dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call"


Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463759).

jerryzh168 commented on torch/ao/quantization/_dbr/quantization_state.py:
```diff
@@ -157,7 +157,9 @@ def reset_to_new_call(self):
-        self.idx = 0
+        # torch.nn.Module __setattr__ has overhead,
+        # this code is the explicit fast path for `self.idx = 0`
+        object.__setattr__(self, 'idx', 0)
```
nit: same comment as before, maybe we can have a set of utility functions for these things

vkuzo replied:
added a comment on the reasoning, not sure if adding one more function would increase the overhead, hope that's ok!

vkuzo committed:
Update on "dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call"


Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

Differential Revision: [D32463759](https://our.internmc.facebook.com/intern/diff/D32463759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463759).

vkuzo committed:
Update on "dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call"


Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

Differential Revision: [D32463759](https://our.internmc.facebook.com/intern/diff/D32463759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463759).

vkuzo committed:
Update on "dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call"


Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

Differential Revision: [D32463759](https://our.internmc.facebook.com/intern/diff/D32463759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463759).

vkuzo committed:
Update on "dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call"


Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

Differential Revision: [D32463759](https://our.internmc.facebook.com/intern/diff/D32463759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463759).

vkuzo committed:
Update on "dbr quant overhead[7/x]: speed up AutoQuantizationState.reset_to_new_call"


Summary:

Speeds up `AutoQuantizationState.reset_to_new_call` by going around
the getattr and setattr overhead in `torch.nn.Module`.

Test plan:

```
// MobileNetV2, 1x3x224x224 input, % of time spent by function during DBR convert

// before
reset_to_new_call - 1.09%

// after
reset_to_new_call - 0.18%
```

Differential Revision: [D32463759](https://our.internmc.facebook.com/intern/diff/D32463759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32463759).
